src/GafferSceneUI/LightTool.cpp

```diff
@@ -0,0 +1,1841 @@
+				}
+
+				glPopMatrix();
+				glPopAttrib();
+			}
+		}
+
+	private :
+
+		void dragBegin( const DragDropEvent &event ) override
+		{
+			const auto &[coneAngle, penumbraAngle] = spotLightAngleValues();
+
+			m_dragStartAngle = handleAngle(
+				m_handleType == HandleType::Cone ? coneAngle : penumbraAngle.value()
+			);
+
+			m_drag = AngularDrag(
+				this,
+				V3f( 0, 0, 0 ),
+				V3f( 0, 1.f, 0 ),
+				V3f( 0, 0, -1.f ),
+				event
+			);
+
+			if( getLookThroughLight() )
+			{
+				const Line3f clickLine( event.line.p0, event.line.p1 );
+				const Line3f originLine( V3f( 0 ), V3f( 0, 0, -1.f ) );
+				const Line3f handleLine(
+					V3f( 0 ),
+					V3f( 0, 0, -1.f ) * M44f().rotate( V3f( 0, degreesToRadians( m_dragStartAngle ), 0 ) )
+				);
+
+				const float clickRaster = rasterDragDistance( clickLine );
+				const float originRaster = rasterDragDistance( originLine );
+				const float handleRaster = rasterDragDistance( handleLine );
+
+				const float delta = handleRaster - originRaster;
+
+				m_rasterXOffset = delta - clickRaster;
+				m_rasterZPosition = abs( delta ) / tan( degreesToRadians( m_dragStartAngle ) );
+			}
+		}
+
+		std::pair<float, std::optional<float>> spotLightAngleValues() const
+		{
+			ScenePlug::PathScope pathScope( handleScenePath()->getContext() );
+			pathScope.setPath( &handleScenePath()->names() );
+
+			float coneAngle = 0;
+
+			if( Inspector::ResultPtr coneInspection = m_coneAngleInspector->inspect() )
+			{
+				if( auto coneAngleData = runTimeCast<const FloatData>( coneInspection->value() ) )
+				{
+					coneAngle = coneAngleData->readable();
+				}
+			}
+
+			std::optional<float> penumbraAngle = std::nullopt;
+
+			if( Inspector::ResultPtr penumbraInspection = m_penumbraAngleInspector ? m_penumbraAngleInspector->inspect() : nullptr )
+			{
+				if( auto penumbraAngleData = runTimeCast<const FloatData>( penumbraInspection->value() ) )
+				{
+					penumbraAngle = penumbraAngleData->readable();
+				}
+			}
+
+			return {coneAngle, penumbraAngle};
+		}
+
+		float handleAngle(const float angle ) const
+		{
+			if( m_handleType == HandleType::Penumbra && m_penumbraType != g_absolutePenumbraType )
+			{
+				return angle;
+			}
+
+			return angle * 0.5f;
+		}
+
+		// Find the intersection of a line in gadget space with a unit sphere and
+		// project that intersection to the handle's plane in raster space. Return
+		// the projected point's distance from the raster center.
+		float rasterDragDistance( const Line3f &ray )
+		{
+			V3f sphereIntersection;
+			Sphere3f( V3f( 0 ), 1.f ).intersect( ray, sphereIntersection );
+
+			const V2f gadgetRasterOrigin = m_view->viewportGadget()->gadgetToRasterSpace( V3f( 0, 0, -1.f ), this );
+			const V2f rasterSphereIntersection = m_view->viewportGadget()->gadgetToRasterSpace( sphereIntersection, this );
+			const V2f rasterNormal = ( m_view->viewportGadget()->gadgetToRasterSpace( V3f( 0, 1.f, -1.f ), this ) - gadgetRasterOrigin ).normalized();
```

Comment:
It would be more intuitive if we could do `gadgetToRasterSpace( V3f( 0, 1, 0 ) )` but `gadgetToRasterSpace` divides by the (negative) z coordinate for perspective projection. So using `-1` for Z here compensates for that transform.

Perhaps it needs a better name, or a comment explaining it?

Reply:
Looking at this again ... I think it's fine. We're subtracting from gadgetRasterOrigin that also has the -1 Z value, that totally makes sense and I'm not sure why I missed this before.